internal/blocktx/processor.go

```diff
@@ -658,56 +738,205 @@ func (p *Processor) markTransactionsAsMined(ctx context.Context, blockID uint64,
 			}
 		}
 	}
 
 	tracing.EndTracing(iterateMerkleTree)
 
 	// update all remaining transactions
-	updateResp, err := p.store.UpsertBlockTransactions(ctx, blockID, txs)
+	err = p.store.UpsertBlockTransactions(ctx, blockID, txs)
 	if err != nil {
-		return errors.Join(ErrFailedToInsertBlockTransactions, fmt.Errorf("block height: %d", blockHeight), err)
+		return errors.Join(ErrFailedToInsertBlockTransactions, fmt.Errorf("block height: %d", block.Height), err)
 	}
 
-	for _, updResp := range updateResp {
-		txBlock := &blocktx_api.TransactionBlock{
-			TransactionHash: updResp.Hash[:],
-			BlockHash:       blockhash[:],
-			BlockHeight:     blockHeight,
-			MerklePath:      updResp.MerklePath,
+	return nil
+}
+
+func (p *Processor) updateOrphans(ctx context.Context, incomingBlock *blocktx_api.Block, competing bool) (chain, bool, error) {
+	chain := []*blocktx_api.Block{incomingBlock}
+
+	uow, err := p.store.StartUnitOfWork(ctx)
+	if err != nil {
+		return nil, false, err
+	}
+	defer func() {
+		_ = uow.Rollback()
+	}()
+
+	// Very important step, this will lock blocks
+	// table for writing but still allow reading.
+	err = uow.WriteLockBlocksTable(ctx)
+	if err != nil {
+		return nil, false, err
+	}
+
+	orphanedBlocks, err := uow.GetOrphanedChainUpFromHash(ctx, incomingBlock.Hash)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(orphanedBlocks) == 0 {
+		return chain, competing, nil
+	}
+
+	blockStatusUpdates := make([]store.BlockStatusUpdate, len(orphanedBlocks))
+	for i := range orphanedBlocks {
+		// We want to mark all orphaned blocks as STALE
+		// in case there already exists a block at any
+		// of their height with status LONGEST, which
+		// would cause constraint validation (height, is_longest).
+		//
+		// If they are part of the LONGEST chain, the reorg
+		// will happen and update their statuses accordingly.
+		orphanedBlocks[i].Status = blocktx_api.Status_STALE
+
+		blockStatusUpdates[i] = store.BlockStatusUpdate{
+			Hash:   orphanedBlocks[i].Hash,
+			Status: blocktx_api.Status_STALE,
 		}
-		err = p.mqClient.PublishMarshal(MinedTxsTopic, txBlock)
-		if err != nil {
-			p.logger.Error("failed to publish mined txs", slog.String("hash", blockhash.String()), slog.Uint64("height", blockHeight), slog.String("err", err.Error()))
+	}
+
+	err = uow.UpdateBlocksStatuses(ctx, blockStatusUpdates)
+	if err != nil {
+		return nil, false, err
+	}
+
+	err = uow.Commit()
+	if err != nil {
+		return nil, false, err
+	}
+
+	p.logger.Info("orphans were found and updated", slog.Int("len", len(orphanedBlocks)))
+
+	chain = append(chain, orphanedBlocks...)
+
+	// if we found any orphans and marked them as STALE
+	// we need to find out if they are part of the longest
+	// or stale chain, so competing is returned as true
+	return chain, true, nil
+}
+
+func (p *Processor) performReorg(ctx context.Context, staleChainTip *blocktx_api.Block) ([]store.TransactionBlock, error) {
+	uow, err := p.store.StartUnitOfWork(ctx)
```

Comment:
I think a trace `performReorg` could be useful

Reply:
Right, added tracing in multiple places